### PR TITLE
[HW05] Add tests for HW02 and HW03

### PR DESCRIPTION
**Pull request by Nij4life**

### CODE REVIEW

File: HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_03_spec.rb [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        2018-04-23 17:17:49.7 ubuntu-xenial[14319] Debug - connecting to: 10.6.246.101
+        2018-04-23 17:17:49.8 ubuntu-xenial[14319] Debug - docker event processed
+        2018-04-23 17:18:19.5 ubuntu-xenial[14319] Debug - monitoring grid communication health
+        2018-04-23 17:18:38.8 ubuntu-xenial[14319] Debug - connecting to: 10.6.246.101
+        2018-04-23 17:18:59.8 ubuntu-xenial[14319] Debug - inside docker_handle_event
+      LOGS
+    end
+
+    it "return '0'" do
+      result = task3(logs)
```
Comment (AliakseiAndryianau) [inline HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_03_spec.rb:54]: please do not use local variables

File: HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_01_spec.rb [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    let(:logs) do
+      <<~LOGS
+        10.6.246.103 - - [23/Apr/2018:20:30:39 +0300] "POST /grid/2/messages HTTP/1.1" 200 48 0.0498
+        10.6.246.101 - - [23/Apr/2018:20:30:42 +0300] "POST /grid/2/event HTTP/1.1" 200 - 0.2277
+        10.6.246.101 - - [23/Apr/2018:20:29:39 +0300] "POST /grid/2/messages HTTP/1.1" 200 48 0.0290
+      LOGS
+    end
+
+    it 'returns empty string' do
+      result = task1(logs)
```
Comment (AliakseiAndryianau) [inline HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_01_spec.rb:30]: please do not use local variables here and further

Comment (AliakseiAndryianau) [reply HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_01_spec.rb:30]: take a look at this article please: https://mixandgo.com/learn/let-vs-instance-variables-in-rspec

File: HW02/aleksandr_serikov/array_refinements_spec.rb [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+      end
+    end
+
+    context 'case block given' do
+      it 'selects objects with a suitable condition' do
+        expect(arr.select { |e| e.odd? }).to eql([1, 3, 5])
+      end
+
+      it 'not modified someself' do
+        arr.each { |e| e * 2 }
```
[the remark this answers is not in the document]
Comment (Nij4life) [reply HW02/aleksandr_serikov/array_refinements_spec.rb:50]: Thanks.
Done

File: HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_03_spec.rb (hunk 1 above)
Comment (Nij4life) [reply HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_03_spec.rb:54]: Done

File: HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_01_spec.rb (hunk 2 above)
Comment (Nij4life) [reply HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_01_spec.rb:30]: Done

File: HW02/aleksandr_serikov/array_refinements_spec.rb [hunk 4]
```diff
@@ -0,0 +1,54 @@
+require_relative 'array_refinements'
+
+using ArrayRefinement
+
+# Testing module ArrayRefinement refined class Array methods
+RSpec.describe ArrayRefinement do
+  let(:arr) { [1, 2, 3, 4, 5] }
+
+  describe '.each' do
+    context 'case no block given' do
```
Comment (B0RlS) [inline HW02/aleksandr_serikov/array_refinements_spec.rb:10]:
```suggestion
    context 'when block is not given' do
```

File: HW02/aleksandr_serikov/array_refinements_spec.rb [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+      it 'returns nil' do
+        expect(arr.each).to eql(nil)
+      end
+    end
+
+    context 'case block given' do
```
Comment (B0RlS) [inline HW02/aleksandr_serikov/array_refinements_spec.rb:16]: same

File: HW02/aleksandr_serikov/array_refinements_spec.rb [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        expect(arr.select).to eql(nil)
+      end
+    end
+
+    context 'case block given' do
+      it 'selects objects with a suitable condition' do
+        expect(arr.select { |e| e.odd? }).to eql([1, 3, 5])
+      end
+
+      it 'result and initial object are different' do
```
Comment (B0RlS) [inline HW02/aleksandr_serikov/array_refinements_spec.rb:49]:
```suggestion
      it 'returns different....' do
```

File: HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_01_spec.rb [hunk 7]
```diff
@@ -0,0 +1,32 @@
+require_relative '../a_serikov_hw_03_t_01'
+
+describe 'task1' do
+  context 'we have an error line' do
```
Comment (B0RlS) [inline HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_01_spec.rb:4]:
```suggestion
  context 'when logs have line with error' do
```

File: HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_02_spec.rb [hunk 8]
```diff
@@ -0,0 +1,36 @@
+require_relative '../a_serikov_hw_03_t_02'
+
+describe 'task2' do
+  context 'String that containing information about post requests' do
```
Comment (B0RlS) [inline HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_02_spec.rb:4]:
```suggestion
  context 'when logs have information about post requests' do
```

File: HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_02_spec.rb [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+         10.6.246.101 - - [23/Apr/2018:20:31:39 +0300] "POST /test/2/messages HTTP/1.1" 200 48 0.0290
+      LOGS
+    end
+
+    it 'returns an array of formatted strings' do
+      expected_result = [
+        "23/Apr/2018:20:30:39 +0300 FROM: 10.6.246.103 TO: /TEST/2/MESSAGES",
+        "23/Apr/2018:20:30:42 +0300 FROM: 10.6.246.101 TO: /TEST/2/RUN",
+        "23/Apr/2018:20:31:39 +0300 FROM: 10.6.246.101 TO: /TEST/2/MESSAGES"
+      ]
```
Comment (B0RlS) [inline HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_02_spec.rb:19]: move this expected_result to `let`

File: HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_03_spec.rb [hunk 10]
```diff
@@ -0,0 +1,57 @@
+require_relative '../a_serikov_hw_03_t_03'
+
+describe 'task3' do
+  context 'log has two events with call core' do
```
Comment (B0RlS) [inline HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_03_spec.rb:4]:
```suggestion
  context 'when log has two core events' do
```

File: HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_03_spec.rb [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+        2018-04-23 17:17:49.7 ubuntu-xenial[14319] Debug - connecting to: 10.6.246.101
+        2018-04-23 17:17:49.8 ubuntu-xenial[14319] Debug - docker event processed
+        2018-04-23 17:18:19.5 ubuntu-xenial[14319] Debug - monitoring grid communication health
+        2018-04-23 17:18:38.8 ubuntu-xenial[14319] Debug - Calling core with action: messages
+        2018-04-23 17:18:38.8 ubuntu-xenial[14319] Debug - connecting to: 10.6.246.101
+        2018-04-23 17:18:59.8 ubuntu-xenial[14319] Debug - inside docker_handle_event
+      LOGS
+    end
+
+    it 'return a string duration of the action in SECONDS between core call events' do
```
Comment (B0RlS) [inline HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_03_spec.rb:17]:
```suggestion
    it 'returns the period between core events' do
```

File: HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_03_spec.rb [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+      expect(task3(logs)).to eql('49.1')
+    end
+  end
+
+  context 'log has four events with call core' do
```
Comment (B0RlS) [inline HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_03_spec.rb:22]: same

File: HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_03_spec.rb [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+        2018-04-23 17:18:19.5 ubuntu-xenial[14319] Debug - monitoring grid communication health
+        2018-04-23 17:18:38.8 ubuntu-xenial[14319] Debug - Calling core with action: messages
+        2018-04-23 17:18:38.8 ubuntu-xenial[14319] Debug - connecting to: 10.6.246.101
+        2018-04-23 17:18:59.8 ubuntu-xenial[14319] Debug - Calling core with action: event2
+        2018-04-23 17:20:00.8 ubuntu-xenial[14319] Debug - inside docker_handle_event
+        2018-04-23 17:21:38.8 ubuntu-xenial[14319] Debug - Calling core with action: messages2
+      LOGS
+    end
+
+    it 'return a string duration of the action in SECONDS between core call events' do
```
Comment (B0RlS) [inline HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_03_spec.rb:37]: same

File: HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_04_spec.rb [hunk 14]
```diff
@@ -0,0 +1,10 @@
+require_relative '../a_serikov_hw_03_t_04'
+
+describe 'task4' do
+  context 'return the count of english letters and digits in that string' do
```
Comment (B0RlS) [inline HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_04_spec.rb:4]:
```suggestion
  context 'when .....' do
```

File: HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_04_spec.rb [hunk 15]
```diff
@@ -0,0 +1,10 @@
...
+    it { expect(task4("hel2!lo")).to eql(letters: 5, digits: 1) }
+    it { expect(task4("wicked .. !")).to eql(letters: 6, digits: 0) }
+    it { expect(task4("абракадабра")).to eql(letters: 0, digits: 0) }
```
Comment (B0RlS) [inline HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_04_spec.rb:7]:
```suggestion
    it { expect(task4("абракадабра")).to eql(letters: 0, digits: 0) }
```
move it to another context

File: HW02/aleksandr_serikov/array_refinements_spec.rb (hunk 4 above)
Comment (Nij4life) [reply HW02/aleksandr_serikov/array_refinements_spec.rb:10]: Done

File: HW02/aleksandr_serikov/array_refinements_spec.rb (hunk 5 above)
Comment (Nij4life) [reply HW02/aleksandr_serikov/array_refinements_spec.rb:16]: Done

File: HW02/aleksandr_serikov/array_refinements_spec.rb (hunk 6 above)
Comment (Nij4life) [reply HW02/aleksandr_serikov/array_refinements_spec.rb:49]: replaced and corrected the test

File: HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_01_spec.rb (hunk 7 above)
Comment (Nij4life) [reply HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_01_spec.rb:4]: Done

File: HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_02_spec.rb (hunk 8 above)
Comment (Nij4life) [reply HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_02_spec.rb:4]: Done

File: HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_02_spec.rb (hunk 9 above)
Comment (Nij4life) [reply HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_02_spec.rb:19]: done

File: HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_03_spec.rb (hunk 10 above)
Comment (Nij4life) [reply HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_03_spec.rb:4]: Done

File: HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_03_spec.rb (hunk 11 above)
Comment (Nij4life) [reply HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_03_spec.rb:17]: Done

File: HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_03_spec.rb (hunk 13 above)
Comment (Nij4life) [reply HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_03_spec.rb:37]: Done

File: HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_04_spec.rb (hunk 15 above)
Comment (Nij4life) [reply HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_04_spec.rb:7]: Done

File: HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_04_spec.rb (hunk 14 above)
Comment (Nij4life) [reply HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_04_spec.rb]: Done

File: HW02/aleksandr_serikov/array_refinements_spec.rb [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+    context 'when block is not given' do
+      it 'returns nil' do
+        expect(arr.each).to eql(nil)
+      end
+    end
+
+    context 'when block is given' do
+      it 'not modified' do
+        expect(arr.each { |e| e * 2 }).to eql([1, 2, 3, 4, 5])
+      end
```
Comment (B0RlS) [inline HW02/aleksandr_serikov/array_refinements_spec.rb:19]:
```suggestion
    context 'when block is given' do
      let(:expected_array) { [1, 2, 3, 4, 5] }
      
      it 'does not modify the existing array' do
        expect(arr.each { |e| e * 2 }).to eql(expected_array)
      end
```

File: HW02/aleksandr_serikov/array_refinements_spec.rb [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+
+  describe '.map' do
+    context 'when block is not given' do
+      it 'returns nil' do
+        expect(arr.map).to eql(nil)
+      end
+    end
+
+    context 'when block is given' do
+      it 'modified' do
```
Comment (B0RlS) [inline HW02/aleksandr_serikov/array_refinements_spec.rb:31]:
```suggestion
      it 'returns modified array' do
```

File: HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_01_spec.rb [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+      <<~LOGS
+        10.6.246.103 - - [23/Apr/2018:20:30:39 +0300] "POST /grid/2/messages HTTP/1.1" 200 48 0.0498
+        10.6.246.101 - - [23/Apr/2018:20:30:42 +0300] "POST /grid/2/event HTTP/1.1" 200 - 0.2277
+        2018-04-23 20:30:42: SSL error, peer: 10.6.246.101, peer cert: , #<Puma::MiniSSL::SSLError: System error: Undefined error: 0 - 0>
+        10.6.246.101 - - [23/Apr/2018:20:29:39 +0300] "POST /grid/2/messages HTTP/1.1" 200 48 0.0290
+      LOGS
+    end
+
+    it 'returns string with error' do
+      expect(task1(logs)).to eql('2018-04-23 20:30:42: SSL error, peer: 10.6.246.101, peer cert: , #<Puma::MiniSSL::SSLError: System error: Undefined error: 0 - 0>')
```
Comment (B0RlS) [inline HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_01_spec.rb:15]:
```suggestion
      expect(task1(logs)).to eql(expected_result)
```
move it to let

File: HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_03_spec.rb [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+        2018-04-23 17:21:38.8 ubuntu-xenial[14319] Debug - Calling core with action: messages2
+      LOGS
+    end
+
+    it 'returns the period between core events' do
+      expect(task3(logs)).to eql(["49.1", "21.0", "159.0"])
+    end
+  end
+
+  context 'there is no event with the call core in the log' do
```
Comment (B0RlS) [inline HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_03_spec.rb:42]:
```suggestion
  context 'when there are no core events in the logs' do
```

File: HW02/aleksandr_serikov/array_refinements_spec.rb (hunk 16 above)
Comment (Nij4life) [reply HW02/aleksandr_serikov/array_refinements_spec.rb:19]: Done

File: HW02/aleksandr_serikov/array_refinements_spec.rb [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+        expect(arr.select).to eql(nil)
+      end
+    end
+
+    context 'when block is given' do
+      it 'selects objects with a suitable condition' do
+        expect(arr.select { |e| e.odd? }).to eql([1, 3, 5])
+      end
+
+      it 'return different object' do
```
[the remark this answers is not in the document]
Comment (Nij4life) [reply HW02/aleksandr_serikov/array_refinements_spec.rb:49]: Done

File: HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_01_spec.rb (hunk 18 above)
Comment (Nij4life) [reply HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_01_spec.rb:15]: Done

File: HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_03_spec.rb (hunk 19 above)
Comment (Nij4life) [reply HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_03_spec.rb:42]: Done

File: HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_03_spec.rb [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+      <<~LOGS
+        2018-04-23 17:17:49.7 ubuntu-xenial[14319] Debug - connecting to: 10.6.246.101
+        2018-04-23 17:17:49.8 ubuntu-xenial[14319] Debug - docker event processed
+        2018-04-23 17:18:19.5 ubuntu-xenial[14319] Debug - monitoring grid communication health
+        2018-04-23 17:18:38.8 ubuntu-xenial[14319] Debug - connecting to: 10.6.246.101
+        2018-04-23 17:18:59.8 ubuntu-xenial[14319] Debug - inside docker_handle_event
+      LOGS
+    end
+
+    it "return '0'" do
```
[the remark this answers is not in the document]
Comment (Nij4life) [reply HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_03_spec.rb:53]: Done

File: HW02/aleksandr_serikov/array_refinements_spec.rb (hunk 17 above)
Comment (Nij4life) [reply HW02/aleksandr_serikov/array_refinements_spec.rb]: Done

File: HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_04_spec.rb [hunk 22]
```diff
@@ -0,0 +1,23 @@
+require_relative '../a_serikov_hw_03_t_04'
+
+describe 'task4' do
```
Comment (B0RlS) [inline HW03/aleksandr_serikov/spec/a_serikov_hw_03_t_04_spec.rb:3]:
```suggestion
describe '#task4' do
```

**Pull request merged**